src/Microsoft.OpenApi/Validations/Visitors/VisitorBase.cs

```diff
@@ -0,0 +1,52 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. 
+
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.OpenApi.Interfaces;
+
+namespace Microsoft.OpenApi.Validations.Visitors
+{
+    /// <summary>
+    /// The base visitor class for Open Api elements.
+    /// </summary>
+    internal abstract class VisitorBase<T> : IVisitor where T : IOpenApiElement
+    {
```

Comment:
This class has no state and we have an IVisitor interface. Is there any reason these two methods couldn't just be library functions?  I don't think we need a base class for all the visitors.

Reply:
Each visitor has two parts:

1. Visit itself's property
2. Go to visit it's Children.


Where:
a. Visit() method is for the #1, it has the same logic for all elements. So, put its at the base class.

b. Next() method is for the #2, it has different logic for all elements. So, each element visitor class should override Next() method to list the children. 

---
In reply to: [157326815](https://github.com/Microsoft/OpenAPI.NET/pull/152#discussion_r157326815) [](ancestors = 157326815)

Reply:
Part of the reason I initially used the visitor pattern was so that we could define the traversal of the OpenAPI DOM just once and then have different visitors for different purposes.  Validation is the obvious Visitor, but code generation could be another.  Either client code gen, or server code scaffolding.  
The Visitor related classes in this PR are specifically tied to validation via the ValidationContext.  Even the abstract base class and IVisitor interface.

If you look at the Visitor pattern https://en.wikipedia.org/wiki/Visitor_pattern here, the logic to walk the object graph is not in the Visitor, it is in the object graph itself.  Rather than embed the Accept methods into our DOM, I created the walker to house all of the traversal logic.

Also, you will see the Visitor implements all the visit methods for all the elements it is interested in.  This allows the Visitor to hold shared state whilst doing the traversal.  It also plays the role of the context object.

With the current approach, ignoring the fact that it is tied to validation, each new "visiting" capability needs to create classes for each and every element and then pass the context between them.

I'm going to see if I can use my walker and validationVisitor to execute your rules classes and we can compare the end result.

Reply:
Sure!

---
In reply to: [157663995](https://github.com/Microsoft/OpenAPI.NET/pull/152#discussion_r157663995) [](ancestors = 157663995)

Reply:
I was able to incorporate your Rules classes into the OpenApiValidator like this,

```C#
   public class OpenApiValidator : OpenApiVisitorBase
    {
        readonly ValidationRuleSet _ruleSet;
        readonly ValidationContext _context;

        /// <summary>
        /// Create a vistor that will validate an OpenAPIDocument
        /// </summary>
        /// <param name="ruleSet"></param>
        public OpenApiValidator(ValidationRuleSet ruleSet = null)
        {
            _ruleSet = ruleSet ?? ValidationRuleSet.DefaultRuleSet;
            _context = new ValidationContext(_ruleSet);
        }

        public override void Visit(OpenApiDocument item) => Validate(item);
        public override void Visit(OpenApiInfo item) => Validate(item);
        public override void Visit(OpenApiContact item) => Validate(item);
        public override void Visit(OpenApiResponse item) => Validate(item);


        public IEnumerable<ValidationError> Errors => _context.Errors;

        private void Validate<T>(T item)
        {
            if (item == null) return;  // Required fields should be checked by higher level objects
            var rules = _ruleSet.Where(r => r.ElementType == typeof(T));
            foreach (var rule in rules)
            {
                rule.Evaluate(_context, item);
            }
        }
    }
}
```
Obviously this doesn't have the Visit methods for all types yet, and some types may be more complex than just the simple expression.  Also, the walker class would need to build the property path.  
However, using this approach, we need one class for walking the DOM (we could create other walkers if other visitors need to traverse in a different way) and one class for dispatching the Visit methods to the rules.  This makes implementing other types of Visitors easier by only requiring one new class.  
My understanding is that this approach would remove the need for : 
- the Visitor class for each element
- the OpenApiVisitorSet
- the ValidationContextExtensions

I'm still not quite sure how the ValidationMap and ValidateCollection extensions fit into this approach.

Reply:
It's ok. Please share your thoughts and codes. We can discuss to select one or just merge together. 

---
In reply to: [157841020](https://github.com/Microsoft/OpenAPI.NET/pull/152#discussion_r157841020) [](ancestors = 157841020)